benchmarks/benchmark_gpu_sparsity.py

```diff
@@ -0,0 +1,248 @@
+import argparse
+import random
+
+import pandas as pd
+import torch
+import torch.utils.benchmark as benchmark
+import torch.nn.functional as F
+from torch import nn
+from torch.sparse import SparseSemiStructuredTensor, to_sparse_semi_structured
+from tqdm import tqdm
+
+
+torch.set_printoptions(
+    precision=2,
+    threshold=None,
+    edgeitems=16,
+    linewidth=480,
+    profile=None,
+    sci_mode=False,
+)
+
+def create_block_sparse_tensor(M, N, blocksize, sparsity, dtype):
+    assert sparsity <= 1.0 and sparsity >= 0.0, \
+        "sparsity should be a value between 0 and 1"
+    A = torch.bernoulli(torch.full((M//blocksize, N//blocksize),
+                        1 - sparsity, dtype=dtype))
+    A = torch.repeat_interleave(A, blocksize, dim=0)
+    A = torch.repeat_interleave(A, blocksize, dim=1)
+    return A.to(dtype).contiguous().cuda()
+
+def create_semi_structured_tensor(
+    r, c, dtype
+):
+    """
+    This function returns a 1:2 sparse matrix of size (r, c).
+    Note that this means this matrix will also be 2:4 and 4:8 sparse as well.
+    """
+
+    choices = [[0, 1], [1, 0]]
+    mask_entries = [random.choice(choices) for i in range(r * c // 2)]
+
+    mask = (
+        torch.tensor(mask_entries, dtype=dtype)
+        .reshape(r, c)
+        .contiguous()
+    ).cuda()
+    sparse_weight = torch.rand(r, c).to(dtype).cuda() * mask
+    return sparse_weight
+
+
+def benchmark_in_us(f, *args, **kwargs):
+    t0 = benchmark.Timer(
+        stmt="f(*args, **kwargs)",
+        globals={"args": args, "kwargs": kwargs, "f": f}
+    )
+    return t0.blocked_autorange()
+
+
+def run_gpu_sparse_benchmark(m, k, n, args):
+    dtype = DTYPE_LOOKUP[args.dtype]
+
+    x = torch.randn(n, k).to(dtype).cuda()
+    b = torch.randn(m, dtype=dtype).cuda()
+
+
+    if args.sparsity == "semi-structured":
+        SparseSemiStructuredTensor._FORCE_CUTLASS = args.backend == "cutlass"
+        A = create_semi_structured_tensor(m, k, dtype)
+        A_sparse = to_sparse_semi_structured(A)
+
+    elif args.sparsity == "block-sparse":
+        A = create_block_sparse_tensor(m, k, args.block_size, args.sparsity_level, dtype)
+        A_sparse = A.to_sparse_bsr(blocksize=args.block_size)
+    else:
+        raise ValueError(f"Unknown sparsity: {args.sparsity}")
+
+    if args.eval_fn == "linear":
+        dense_output = F.linear(x, A, b)
+        sparse_output = F.linear(x, A_sparse, b)
+        correct = torch.allclose(dense_output, sparse_output, rtol=1e-3, atol=1e-3)
+        dense_time = benchmark_in_us(F.linear, x, A, b)
+        sparse_time = benchmark_in_us(F.linear, x, A_sparse, b)
+
+    elif args.eval_fn == "mm":
+        dense_output = torch.mm(x, A)
+        sparse_output = torch.mm(x, A_sparse)
+        correct = torch.allclose(dense_output, sparse_output, rtol=1e-3, atol=1e-3)
```

Comment:
Alright, so we do care about correctness. It seems like maybe something to turn on/off. Morally this should be covered by unit tests, but I also never mind more sources of verification.

Reply:
You can use torch.testing.assert_allclose to have it raise an exception with location and error.

Reply:
I think when I first wrote this script, we didn't have tests. Let's just remove this correctness checking - since we have better testing now.